### PR TITLE
Wire the partitioner into the main compilation flow.

### PR DESCRIPTION
**Pull request by stellaraccident**
* Just "partitions" with an arity of 1 (optimized but does not partitioning).
* Several fixups to the partitioner to make it the same API as IREE.
* Mostly forked code from the main compilation flow.
* Adds a github action test to do the smoke test with the partitioner enabled (this is not robust but here pending a better testing strategy).

This is still a bit of a toy and needs to be driven based on execution options to set the number of partitions. It also needs to be extended to work with stablehlo "portable artifacts" (bitcode) for true version stability.

Set the IREE_PJRT_SAVE_ARTIFACTS var and manually inspected that the partitioner was feeding the compiler.

Progress on #82.

### CODE REVIEW

File: iree/integrations/pjrt/common/dylib_platform.cc [hunk 1]
```diff
@@ -41,6 +44,26 @@ std::optional<std::string> LoadCompilerStubOnce(
...
 
+std::optional<std::string> InitializePartitionerForProcess(
+    const std::string& library_path) {
+  if (!openxlaPartitionerLoadLibrary(library_path.c_str())) {
+    return {};
+  }
+
+  openxlaPartitionerGlobalInitialize();
+
+  return library_path;
```
Comment (okkwon) [inline iree/integrations/pjrt/common/dylib_platform.cc:55]: Do we ever want to return a different string than the input `library_path`?
If no, we can convert the function to return a bool.

File: iree/integrations/pjrt/common/dylib_platform.cc [hunk 2]
```diff
@@ -178,4 +228,26 @@ std::optional<std::string> DylibPlatform::GetCompilerLibraryPath() {
...
+  auto found_explicit = config_vars().Lookup("PARTITIONER_LIB_PATH");
+  if (found_explicit) {
+    return *found_explicit;
+  }
+
+  // Try to compute from lib dir.
+  auto lib_dir = GetLibraryDir();
+  if (!lib_dir) return {};
+
+  char* path;
```
Comment (okkwon) [inline iree/integrations/pjrt/common/dylib_platform.cc:241]: nit: initialize it by nullptr.

**Review by okkwon — APPROVED** [review]
Left a few minor comments, but looks great. Thanks

File: iree/integrations/pjrt/common/hlo_partitioner.cc [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    return openxlaPartitionerInvocationParseSource(inv_, source);
+  }
+
+  std::unique_ptr<CompilerOutput> CompileStandardPipeline() override {
+    openxla_partitioner_error_t* error;
+    if (!openxlaPartitionerInvocationPipeline(inv_, "gspmd")) {
+      return nullptr;
+    }
+
+    error = openxlaPartitionerOutputOpenMembuffer(&output_);
```
[the remark this answers is not in the document]
Comment (stellaraccident) [reply iree/integrations/pjrt/common/hlo_partitioner.cc:135]: Done here and in iree_compiler.cc.

File: iree/integrations/pjrt/common/dylib_platform.cc (hunk 1 above)
Comment (stellaraccident) [reply iree/integrations/pjrt/common/dylib_platform.cc:55]: Changed it to a bool. Not sure it is better but that's fine.

**Pull request merged**